Typed ->GTS function and filter.byselector

pi-r-p opened the pull request:
## `->GTS` function
Added a few signatures to `->GTS` function, keeping compatibility.

`LIST<ENCODERS>  ->GTS ` is a map of list of gts. (map keys are the same as existing implementation)

`LIST<ENCODERS> { } ->GTS` is a list of gts, typed by the first element in the encoder

`LIST<ENCODERS> {  'DOUBLE' 'classname{label~x.*}' } ->GTS` is a list of gts, typed by the type in the map input keys, when the encoder metadata matches the selector in the map values. If several selectors match, the first corresponding type is applied. This signature is really helpfull after a MVSPLIT operation:

```
{ 1 'temperature' 2 'sensor status' } MVINDEXSPLIT
{ 'DOUBLE'  'temperature{}'  'BOOLEAN' 'sensor status{}'} ->GTS
```

Extended selectors, with classname, labels and attributes are supported : `'classname{label~x.*}{attribute=y}`

List of selectors are also possible: `LIST<ENCODERS> {  'DOUBLE' [ 'classname{label~x.*}' 'otherClassname{}' ] } ->GTS`

By default, the ` {} ->GTS` signature add an extra `.type` label in the output. The label name could be overriden with `label` key in the input:
```
{ 'DOUBLE'  'temperature{}' 
  'BOOLEAN' 'sensor status{}'
  'label' 'other_label'  // could also be NULL, to avoid adding extra labels
} ->GTS
```

## `filter.byselector` function
Added a new filter.byselector function that also implements selector and extended selectors.
```java
  [ $glist [] '~.*{alive~.*}' 'filter.byselector' EVAL ] FILTER  // look into labels and attributes
  [ $glist [] '~.*{}{alive~.*}' 'filter.byselector' EVAL ] FILTER  //only attributes
  [ $glist [] '~.*{alive~.*}{}' 'filter.byselector' EVAL ] FILTER  //only labels
```

hbs commented on warp10/src/main/java/io/warp10/script/functions/TOGTS.java:
```diff
@@ -16,98 +16,265 @@
+
+    if (top instanceof Map) {
+      typeMap = new LinkedHashMap<>();
+      // this is a map to specify type by name, it should contain valid types
```
Change comment, it is a selector

hbs commented on warp10/src/main/java/io/warp10/script/functions/TOGTS.java:
```diff
@@ -16,98 +16,265 @@
+      // this is a map to specify type by name, it should contain valid types
+      for (Map.Entry<Object, Object> entry: ((Map<Object, Object>) top).entrySet()) {
+        if (!(entry.getKey() instanceof String)) {
+          throw new WarpScriptException(getName() + " type MAP input must contains encoder name as keys.");
```
Change error message to reflect actual usage

hbs commented on warp10/src/main/java/io/warp10/script/functions/TOGTS.java:
```diff
@@ -16,98 +16,265 @@
+        }
+        if (entry.getValue() instanceof String) {
+          String t = (String) entry.getValue();
+          if (t.equals("LONG") || t.equals("DOUBLE") || t.equals("BOOLEAN") || t.equals("STRING")) {
```
BINARY should be included as it will be the responsability of the caller to decode said strings using ISO8859-1 instead of UTF-8

hbs commented on warp10/src/main/java/io/warp10/script/functions/TOGTS.java:
```diff
@@ -16,98 +16,265 @@
+          if (t.equals("LONG") || t.equals("DOUBLE") || t.equals("BOOLEAN") || t.equals("STRING")) {
+            typeMap.put((String) entry.getKey(), (String) entry.getValue());
+          } else {
+            throw new WarpScriptException(getName() + " type MAP input must contains valid types as value (LONG, DOUBLE, BOOLEAN or STRING).");
```
missing binary

hbs commented on warp10/src/main/java/io/warp10/script/functions/TOGTS.java:
```diff
@@ -16,98 +16,265 @@
+    if (top instanceof List) {
+      for (Object o: (List) top) {
+        if (!(o instanceof String) && !(o instanceof byte[]) && !(o instanceof GTSEncoder)) {
+          throw new WarpScriptException(getName() + " operates on a string, a byte array, an encoder, or list of thereof.");
```
or a list thereof

hbs commented on warp10/src/main/java/io/warp10/script/functions/TOGTS.java:
```diff
@@ -16,98 +16,265 @@
+    } else if (top instanceof String || top instanceof byte[] || top instanceof GTSEncoder) {
+      decodersInput.add(getDecoderFromObject(top));
+    } else {
+      throw new WarpScriptException(getName() + " operates on a string, a byte array, an encoder, or list of thereof.");
```
or a list thereof

hbs commented on warp10/src/main/java/io/warp10/script/functions/TOGTS.java:
```diff
@@ -16,98 +16,265 @@
+        gts.setMetadata(decoder.getMetadata());
+        String enforcedType = null;
+        for (Entry<String, String> entry: typeMap.entrySet()) {
+          if (metadataMatchSelector(decoder.getMetadata(), entry.getKey())) {
```
This will parse the selector on each loop

hbs commented on warp10/src/main/java/io/warp10/script/functions/TOGTS.java:
```diff
@@ -16,98 +16,265 @@
+   * @return true when metadata's classname, labels and attributes fits the selector.
+   * @throws WarpScriptException
+   */
+  public boolean metadataMatchSelector(Metadata metadata, String selector) throws WarpScriptException {
```
does it need to be public?

hbs commented on warp10/src/test/java/io/warp10/continuum/gts/MetadataSelectorMatcherTest.java:
```diff
@@ -0,0 +1,42 @@
+package io.warp10.continuum.gts;
```
Missing copyright notice

hbs commented on warp10/src/main/java/io/warp10/continuum/gts/MetadataSelectorMatcher.java:
```diff
@@ -0,0 +1,178 @@
+  private final String selector;
+
+  private final Matcher classnamePattern;
+  private final boolean withClassnameSelector;
```
the withXXX variables are not needed, simply check for nullity of the assocaited pattern variables

pi-r-p commented on warp10/src/main/java/io/warp10/continuum/gts/MetadataSelectorMatcher.java:
```diff
@@ -0,0 +1,178 @@
+    // read class selector
+    String classSelector = null;
+    try {
+      classSelector = URLDecoder.decode(m.group("classname"), StandardCharsets.UTF_8.name());
```
As soon as you merge the fix, I will cherry pick it to use WarpURLDecoder

pi-r-p replied:
merged with master, now using io.warp10.WarpURLDecoder

pi-r-p commented on warp10/src/main/java/io/warp10/script/functions/TOGTS.java:
```diff
@@ -16,98 +16,208 @@
+        }
+        while (decoder.next()) {
+          Object value = decoder.getBinaryValue();
+          if (value instanceof byte[]) {
```
indeed... first operation done.  I am still learning about binary support !

pi-r-p commented on warp10/src/main/java/io/warp10/script/functions/TOGTS.java:
```diff
@@ -16,98 +16,208 @@
+        }
+        if (entry.getKey() instanceof String) {
+          String t = (String) entry.getKey();
+          if ("LONG".equals(t) || "DOUBLE".equals(t) || "BOOLEAN".equals(t) || "STRING".equals(t) || "BINARY".equals(t)) {
```
I changed the signature meanwhile:
```
  {
   'DOUBLE' '~temperature.*{}' 
   'LONG' 'door_opened_counter{}' 
   'BOOLEAN' '~.*status.*{}' 
  } ->GTS
```

It is simpler to handle, and there is no more "match the first one", which can be very confusing.

pi-r-p replied:
you can now have lists of selectors by type in the signature.

pi-r-p commented:
- Added list of selectors support for each type.
- Added an extra `.type` label in each output gts
- The extra type label could be overridden in the input map, with another string, or null to avoid adding any new label.

hbs commented on warp10/src/main/java/io/warp10/continuum/gts/MetadataSelectorMatcher.java:
```diff
@@ -0,0 +1,190 @@
+//
+//   Copyright 2019  SenX S.A.S.
```
Copyright notice is no longer valid ;-) Happy New Year!

hbs commented on warp10/src/main/java/io/warp10/continuum/gts/MetadataSelectorMatcher.java:
```diff
@@ -0,0 +1,190 @@
+    if (0 == labelsSelectors.size()) {
+      this.labelsPatterns = null;
+    } else {
+      this.labelsPatterns = new HashMap<>(labelsSelectors.size());
```
Fill the diamond with actual types to make code more understandable

hbs commented on warp10/src/main/java/io/warp10/continuum/gts/MetadataSelectorMatcher.java:
```diff
@@ -0,0 +1,190 @@
+  public boolean MetaDataMatch(Metadata metadata) {
+
+    // first, check classname
+    boolean classnameMatch = (null == this.classnamePattern) || null == metadata.getName()
```
This means that if class is not set then a match will always occur?

pi-r-p replied:
Yes. But it should not happen, as NEWGTS and NEWENCODER always set name to "".
Just a safety...

hbs replied:
given the method is public it should be assumed that it could be called with a Metadata instance whose name is not set, and in this case it should not match

pi-r-p replied:
done, with earlier exit if classname does not match

hbs commented on warp10/src/main/java/io/warp10/continuum/gts/MetadataSelectorMatcher.java:
```diff
@@ -0,0 +1,190 @@
+    Map inputAttributes = metadata.getAttributes();
+    boolean labelAndAttributeMatch = true;
+
+    if (null != this.attributesPatterns) {
```
Make the comment explaining how match is performed a little more verbose

pi-r-p replied:
The doc and the tests are explicit. I add a few lines from the doc in the code.

hbs commented on warp10/src/main/java/io/warp10/script/filter/FilterBySelector.java:
```diff
@@ -0,0 +1,83 @@
+//
+//   Copyright 2019  SenX S.A.S.
```
Fix copyright notice

hbs commented on warp10/src/main/java/io/warp10/script/functions/TOGTS.java:
```diff
@@ -16,98 +16,258 @@
+    String extraLabel = DEFAULT_LABEL_NAME;
+
+    if (top instanceof Map) {
+      typeMap = new LinkedHashMap<>();
```
Add type to diamond

hbs commented on warp10/src/main/java/io/warp10/script/functions/TOGTS.java:
```diff
@@ -16,98 +16,258 @@
+          if ("LONG".equals(t) || "DOUBLE".equals(t) || "BOOLEAN".equals(t) || "STRING".equals(t) || "BINARY".equals(t)) {
+            if (entry.getValue() instanceof String) {
+              // MAP with type->selector
+              ArrayList<MetadataSelectorMatcher> l = new ArrayList<>();
```
Add type to diamond

hbs commented on warp10/src/main/java/io/warp10/script/functions/TOGTS.java:
```diff
@@ -16,98 +16,258 @@
+              typeMap.put((String) entry.getKey(), l);
+            } else if (entry.getValue() instanceof List) {
+              // MAP with type->listOfSelectors
+              ArrayList<MetadataSelectorMatcher> l = new ArrayList<>();
```
Add type to diamon

hbs commented on warp10/src/main/java/io/warp10/script/functions/TOGTS.java:
```diff
@@ -16,98 +16,258 @@
-    Map<String,GeoTimeSerie> series = new HashMap<String,GeoTimeSerie>();
-    
+
+    List<GTSDecoder> decodersInput = new ArrayList<>();
```
Add type to diamond

hbs commented on warp10/src/main/java/io/warp10/script/functions/TOGTS.java:
```diff
@@ -16,98 +16,258 @@
+
+    // if there is no type map, the output is a map of GTS (v2.3.0 signature), or a map of lists of GTS
+    if (null == typeMap) {
+      Map<String, ArrayList<GeoTimeSerie>> result = new HashMap<>();
```
Add type to diamond

hbs commented on warp10/src/main/java/io/warp10/script/functions/TOGTS.java:
```diff
@@ -16,98 +16,258 @@
+    if (null == typeMap) {
+      Map<String, ArrayList<GeoTimeSerie>> result = new HashMap<>();
+      for (GTSDecoder decoder: decodersInput) {
+        Map<String, GeoTimeSerie> series = new HashMap<>();
```
ditto

hbs commented on warp10/src/main/java/io/warp10/script/functions/TOGTS.java:
```diff
@@ -16,98 +16,258 @@
+      //  - if the encoder doesn't fit to any selector, the gts will have the type of the first encountered element in the encoder.
+      // GTSHelper.setValue will try to convert values whenever possible, and
+      // a byte array will be serialized as an ISO-8859-1 string by GTSHelper.setValue
+      ArrayList<GeoTimeSerie> result = new ArrayList<>();
```
type in diamond

hbs commented on warp10/src/main/java/io/warp10/script/functions/TOGTS.java:
```diff
@@ -16,98 +16,258 @@
+        GeoTimeSerie gts = new GeoTimeSerie();
+        gts.setMetadata(decoder.getMetadata());
+        String enforcedType = null;
+        for (Entry<String, ArrayList<MetadataSelectorMatcher>> entry: typeMap.entrySet()) {
```
List<> rather than ArrayList

pi-r-p replied:
It leads to an type incompatibly.

hbs commented on warp10/src/test/java/io/warp10/continuum/gts/MetadataSelectorMatcherTest.java:
```diff
@@ -0,0 +1,88 @@
+//
+//   Copyright 2019  SenX S.A.S.
```
Fix year

ftence commented on warp10/src/main/java/io/warp10/script/functions/TOGTS.java:
```diff
@@ -16,98 +16,258 @@
+      // a byte array will be serialized as an ISO-8859-1 string by GTSHelper.setValue
+      ArrayList<GeoTimeSerie> result = new ArrayList<GeoTimeSerie>();
+      boolean classMatch;
+      boolean labelsMatch;
```
Unused.

ftence commented on warp10/src/test/java/io/warp10/continuum/gts/MetadataSelectorMatcherTest.java:
```diff
@@ -0,0 +1,88 @@
+    Metadata test = new Metadata();
+    test.setName("temperature");
+
+    Map labels = new HashMap();
```
Add `<String, String>`.

ftence commented on warp10/src/test/java/io/warp10/continuum/gts/MetadataSelectorMatcherTest.java:
```diff
@@ -0,0 +1,88 @@
+    labels.put("sensor", "23");
+    test.setLabels(labels);
+
+    Map attributes = new HashMap();
```
Add `<String, String>`.

ftence commented on warp10/src/test/java/io/warp10/continuum/gts/MetadataSelectorMatcherTest.java:
```diff
@@ -0,0 +1,88 @@
+    Metadata testemptyclassname = new Metadata();
+    testemptyclassname.setName("");
+
+    labels = new HashMap();
```
Add `<String, String>`.

ftence commented on warp10/src/test/java/io/warp10/continuum/gts/MetadataSelectorMatcherTest.java:
```diff
@@ -0,0 +1,88 @@
+    labels.put("sensor", "23");
+    testemptyclassname.setLabels(labels);
+
+    attributes = new HashMap();
```
Add `<String, String>`.

pi-r-p commented on warp10/src/test/java/io/warp10/continuum/gts/MetadataSelectorMatcherTest.java:
```diff
@@ -0,0 +1,88 @@
+//
+//   Copyright 2019  SenX S.A.S.
```
I didn't changed the test in 2020...

pi-r-p commented on warp10/src/main/java/io/warp10/continuum/gts/MetadataSelectorMatcher.java:
```diff
@@ -0,0 +1,199 @@
+            String inputLabel = (String) inputLabels.get(ls.getKey());
+            if (null != inputLabel) {
+              // label exists in the input, try to match.
+              labelAndAttributeMatch &= ls.getValue().reset(inputLabel).matches();
```
everywhere for all conditions in fact. legacy of a previous implementation. 
didn't care about intellij warning, I should.

I will also check classnameMatch before looking further.

ftence commented on warp10/src/main/java/io/warp10/script/functions/TOGTS.java:
```diff
@@ -16,98 +16,257 @@
+    String extraLabel = DEFAULT_LABEL_NAME;
+
+    if (top instanceof Map) {
+      typeMap = new LinkedHashMap<String, ArrayList<MetadataSelectorMatcher>>();
```
Why use a `LinkedHashMap` even though the given Map does not guarantee order?

pi-r-p replied:
input *is* a LinkedHashMap

ftence commented on warp10/src/main/java/io/warp10/script/functions/TOGTS.java:
```diff
@@ -16,98 +16,257 @@
+              enforcedType = entry.getKey();
+            }
+            if (null != enforcedType) {
+              break;
```
This only break the for on the list of `MetadataSelectorMatcher`, not for all `MetadataSelectorMatcher` for all types.
If several MetadataSelectorMatcher match, the behavior is quite strange.
I think you should check if `enforcedType` is already set and throw an error if it is the case, to avoid GTS matching several types.
Here is a minimal example for this issue:
```
NEWENCODER
{ 
  'BOOLEAN' '~.*{}'
  'STRING' '~.*{}' 
} 
->GTS
```

pi-r-p replied:
Indeed, I must break the type loop also. 
I do not want to raise an error. Order matters in the input map, that's why I used a linkedHashMap.
If user want a fallback to another type, it could be done with 
```
{ 'LONG' 'foo{unit=42}' 
  'DOUBLE' '~.*{}'
} ->GTS
```

hbs commented on warp10/src/main/java/io/warp10/continuum/gts/MetadataSelectorMatcher.java:
```diff
@@ -0,0 +1,199 @@
+    }
+  }
+
+  public boolean MetaDataMatch(Metadata metadata) {
```
or matches as for Matcher?

ftence replied:
Yes, agreed, `matches` is much better.

pi-r-p commented on warp10/src/main/java/io/warp10/script/functions/TOGTS.java:
```diff
@@ -16,98 +16,257 @@
+              l = "STRING";
+            } else {
+              // empty series, undefined type.
+              l = "EMPTY";
```
- Strict compatibility with previous implementation for "without map"
- UNDEFINED make me think to an error, not to an empty one.

hbs replied:
Empty encoders should not be added as UNDEFINED is not a type that can be specified for the selector. BUT, all types for which there is a selector or list thereof should appear in the resulting map, possibly with an empty list of encoders if none matched, this way it makes the use of the function's result smoother.

pi-r-p replied:
output is a list. This could be done for the list of encoders without map input signature. But as discussed, no.

pi-r-p commented on warp10/src/main/java/io/warp10/script/functions/TOGTS.java:
```diff
@@ -16,98 +16,257 @@
+              l = "EMPTY";
+            }
+          }
+          gts.setLabel(extraLabel, l);
```
What if I am overriding it ? throw an exception and fail, keep the old one ? None of these looks good to me. User can change the label name if any problem.

hbs replied:
Type should be marked UNDEFINED nonetheless, not EMPTY

pi-r-p commented on warp10/src/main/java/io/warp10/script/functions/TOGTS.java:
```diff
@@ -16,98 +16,257 @@
 public class TOGTS extends NamedWarpScriptFunction implements WarpScriptStackFunction {
-  
+
+  private static final String LABEL_NAME_PARAMETER = "label";
```
indeed.

pi-r-p commented on warp10/src/main/java/io/warp10/script/functions/TOGTS.java:
```diff
@@ -16,98 +16,257 @@
+            } else if (GeoTimeSerie.TYPE.DOUBLE == gts.getType()) {
+              l = "DOUBLE";
+            } else if (GeoTimeSerie.TYPE.STRING == gts.getType()) {
+              l = "STRING";
```
done

hbs commented on warp10/src/main/java/io/warp10/continuum/gts/MetadataSelectorMatcher.java:
```diff
@@ -0,0 +1,206 @@
+    }
+
+    // Check labels.
+    // ###### Standard selector : `classname{labelOrAttribute=x}`
```
extra whitespace before ':'

hbs commented on warp10/src/main/java/io/warp10/continuum/gts/MetadataSelectorMatcher.java:
```diff
@@ -0,0 +1,206 @@
+    // ###### Standard selector : `classname{labelOrAttribute=x}`
+    //  - If classname match, `filter.byselector` looks into input labels to check if labelOrAttribute exists and equals x.
+    //  If labelOrAttribute is not found among input labels, it looks into input attributes if the label exists and equals x.
+    // ###### Extended selector : `classname{labelname=x}{attributename=y}` matches if:
```
ditto

hbs commented on warp10/src/main/java/io/warp10/continuum/gts/MetadataSelectorMatcher.java:
```diff
@@ -0,0 +1,206 @@
+    //  - `={}` matches only emtpy classnames, whatever the labels and attributes.
+    //  - `~.*{label=value}{} filter.byselector` is equivalent to `{ 'label' 'value' } filter.bylabels`.
+    //  - `~.*{}{attribute~value} filter.byselector` is equivalent to `{ 'attribute' '~value' } filter.byattr`.
+
```
ditto

hbs commented on warp10/src/main/java/io/warp10/script/filter/FilterBySelector.java:
```diff
@@ -0,0 +1,83 @@
+    public Object apply(WarpScriptStack stack) throws WarpScriptException {
+      Object arg = stack.pop();
+      if (!(arg instanceof String)) {
+        throw new WarpScriptException("selector must be a STRING");
```
Add function name in error message.

hbs commented on warp10/src/main/java/io/warp10/script/functions/TOGTS.java:
```diff
@@ -16,98 +16,268 @@
+
+    } else {
+      // if there is a type map on the stack:
+      //  - the map key is a selector. If the encoder's metadata fits, the gts will have the corresponding type of the first matching selector.
```
Change comment to reflect reality

hbs commented on warp10/src/main/java/io/warp10/script/functions/TOGTS.java:
```diff
@@ -16,98 +16,268 @@
+            mustGuessTypeFromFirstValue = false;
+          }
+        }
+        // also set an extra label with the enforced type, or EMPTY if there is no type enforcement match AND the encoder was empty.
```
"UNDEFINED"

hbs commented on warp10/src/main/java/io/warp10/continuum/gts/MetadataSelectorMatcher.java:
```diff
@@ -0,0 +1,206 @@
+      }
+      this.attributesPatterns = new HashMap<String, Matcher>(attributesSelectors.size());
+      // build label patterns map
+      for (Entry<String, String> l: attributesSelectors.entrySet()) {
```
Naming the variable 'a' instead of 'l' would be less disturbing since we are not iterating over (l)abels but (a)ttributes

hbs commented on warp10/src/main/java/io/warp10/continuum/gts/MetadataSelectorMatcher.java:
```diff
@@ -0,0 +1,206 @@
+      if (null != this.labelsPatterns && null != inputLabels) {
+        for (Entry<String, Matcher> ls: this.labelsPatterns.entrySet()) {
+          String inputLabel = (String) inputLabels.get(ls.getKey());
+          labelAndAttributeMatch = null != inputLabel && ls.getValue().reset(inputLabel).matches();
```
Use parenthesis around null != inputLabel

hbs commented on warp10/src/main/java/io/warp10/continuum/gts/MetadataSelectorMatcher.java:
```diff
@@ -0,0 +1,206 @@
+            break;
+          }
+          String inputAttribute = (String) inputAttributes.get(ls.getKey());
+          labelAndAttributeMatch = null != inputAttribute && ls.getValue().reset(inputAttribute).matches();
```
Use parens around null != inputAttribute

hbs commented on warp10/src/main/java/io/warp10/continuum/gts/MetadataSelectorMatcher.java:
```diff
@@ -0,0 +1,206 @@
+            } else if (null != inputAttributes) {
+              // label does not exist, look for attribute existence and match.
+              String inputAttribute = (String) inputAttributes.get(ls.getKey());
+              labelAndAttributeMatch = null != inputAttribute && ls.getValue().reset(inputAttribute).matches();
```
Use parenthesis around null != inputAttribute

hbs commented on warp10/src/test/java/io/warp10/continuum/gts/MetadataSelectorMatcherTest.java:
```diff
@@ -0,0 +1,89 @@
+public class MetadataSelectorMatcherTest {
+
+  @Test
+  public void metaDataMatch() throws WarpScriptException {
```
rename method testMatches()

pi-r-p replied:
⡎⢱ ⣇⠜
 ⠣⠜ ⠇⠱

ftence commented on warp10/src/main/java/io/warp10/script/functions/TOGTS.java:
```diff
@@ -16,98 +16,268 @@
 /**
  * Converts an encoder into a map of gts, one per type
+ * TODO: Provide a map type->selector or type->list of selectors, with an optional 'label' key. By default, the enforced type will be added in a ".type" label.
+ * label = null will not add any label to the output GTS.
```
Done! You can remove that!

ftence commented on warp10/src/test/java/io/warp10/continuum/gts/MetadataSelectorMatcherTest.java:
```diff
@@ -0,0 +1,87 @@
+
+
+  }
+}
```
Add new line.

pi-r-p replied:
```
 mmmmmmmm  mm    mm  mm    mm  mm    mm  mm    mm 
 ##""""""  ##    ##  ##    ##  ##    ##  ##    ## 
 ##        ##    ##  ##    ##  ##    ##  ##    ## 
 #######   ##    ##  ##    ##  ##    ##  ##    ## 
 ##        ##    ##  ##    ##  ##    ##  ##    ## 
 ##        "##mm##"  "##mm##"  "##mm##"  "##mm##" 
 ""          """"      """"      """"      """"   
 ```

hbs commented:
Congrats!
